crates/tx5/src/lib.rs

```diff
@@ -36,261 +36,177 @@ pub mod deps {
+
+    fn poll(
+        mut self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Self::Output> {
+        std::pin::Pin::new(&mut self.f).poll(cx)
     }
 }
-*/
 
-impl bytes::Buf for BytesList {
-    fn remaining(&self) -> usize {
-        self.0.iter().map(|b| b.remaining()).sum()
-    }
+#[cfg(test)]
+mod test_behavior;
 
-    fn chunk(&self) -> &[u8] {
-        match self.0.front() {
-            Some(b) => b.chunk(),
-            None => &[],
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[tokio::test(flavor = "multi_thread")]
+    async fn atsf_traits() {
+        fn check<F>(_f: F)
+        where
+            F: Send + Sync + Unpin,
+        {
         }
-    }
 
-    #[allow(clippy::comparison_chain)] // clearer written explicitly
-    fn advance(&mut self, mut cnt: usize) {
-        loop {
-            let mut item = match self.0.pop_front() {
-                Some(item) => item,
-                None => return,
-            };
-
-            let rem = item.remaining();
-            if rem == cnt {
-                return;
-            } else if rem < cnt {
-                cnt -= rem;
-            } else if rem > cnt {
-                item.advance(cnt);
-                self.0.push_front(item);
-                return;
-            }
+        let a = AbortableTimedSharedFuture::new(
+            std::time::Duration::from_millis(10),
+            Error::str("my timeout err").into(),
+            async move { Ok(()) },
+        );
+
+        check(a);
+    }
+
+    #[tokio::test(flavor = "multi_thread")]
+    async fn atsf_happy() {
+        AbortableTimedSharedFuture::new(
+            std::time::Duration::from_secs(1),
+            Error::id("to").into(),
+            async move { Ok(()) },
+        )
+        .await
+        .unwrap();
+    }
+
+    #[tokio::test(flavor = "multi_thread")]
+    async fn atsf_timeout() {
+        let r = AbortableTimedSharedFuture::new(
+            std::time::Duration::from_millis(1),
+            Error::id("to").into(),
+            async move {
+                tokio::time::sleep(std::time::Duration::from_secs(1)).await;
+                Ok(())
+            },
+        )
+        .await;
+        assert_eq!("to", r.unwrap_err().to_string());
+    }
+
+    #[tokio::test(flavor = "multi_thread")]
+    async fn atsf_abort() {
+        let a = AbortableTimedSharedFuture::new(
+            std::time::Duration::from_secs(1),
+            Error::id("to").into(),
+            async move {
+                tokio::time::sleep(std::time::Duration::from_secs(1)).await;
+                Ok(())
+            },
+        );
+        {
+            let a = a.clone();
+            tokio::task::spawn(async move {
+                tokio::time::sleep(std::time::Duration::from_millis(10)).await;
+                a.abort(Error::id("abort").into());
+            });
         }
+        let r = a.await;
+        assert_eq!("abort", r.unwrap_err().to_string());
     }
 }
```

Comment:
Is there a way to test the drop behavior? Is it important?

Reply:
Drop logic was removed here: https://github.com/holochain/tx5/pull/77/commits/4a24d8b6c0eff1218834dd8d0dd4971ed051602f